### PR TITLE
8324859: Improve error recovery

### PR DESCRIPTION
**Pull request by lahodaj**
Consider code like:
```
package tests;
public class TestB {
    public static boolean test() // <--- missing open brace
        return true;
    }
    public static boolean test2() {
        return true;
    }
}
```

In JDK 17, it used to produce 3 compile-time errors:
```
$ javac  /tmp/TestB.java 
/tmp/TestB.java:3: error: ';' expected
    public static boolean test() // <--- missing open brace
                                ^
/tmp/TestB.java:6: error: class, interface, enum, or record expected
    public static boolean test2() {
                  ^
/tmp/TestB.java:8: error: class, interface, enum, or record expected
    }
    ^
3 errors
```

Currently, it produces 4:
```
$ javac  /tmp/TestB.java 
/tmp/TestB.java:3: error: ';' expected
    public static boolean test() // <--- missing open brace
                                ^
/tmp/TestB.java:6: error: implicitly declared classes are a preview feature and are disabled by default.
    public static boolean test2() {
                  ^
  (use --enable-preview to enable implicitly declared classes)
/tmp/TestB.java:9: error: class, interface, enum, or record expected
}
^
/tmp/TestB.java:1: error: implicitly declared class should not have package declaration
package tests;
^
4 errors
```

Neither of these is particularly nice. The common property is that the javac's parser "de-synchronizes" on the missing opening brace (`{`), and consumes the first closing brace (`}`) as a closing brace for the class (not for the method), causing all the follow-up behavior.

In general, the javac's parser handles missing closing braces fairly well - it skips tokens until it find something it can synchronize on, and continues. But, it does not handle missing opening braces as well, and has tendency to get lost in the input.

For the case above, it would be much better if javac parsed the code following the missing opening brace as a block.

This patch is attempting to do that, without making other cases I was able to find (much) worse. The overall approach is to skip tokens until a statement or member start is found, and then look at what follows:
- if it is an opening brace, parse the follow up as a block
- if it is a statement keyword (like `if`), parse the follow up as a block,
- if it is a closing brace, see if injecting a virtual opening brace at the current position would balance the opening/closing braces in the rest of the file - if yes, parse the follow up as a block,
- otherwise, try to speculatively parse the follow up - if it looks like a well-formed block, let the main parser parse it as a block.

There is a small trick by checking whether the parameters were parsed OK - and not do most of the above if there was a problem while parsing the parameters. This is to keep the existing behavior in the `testImplicitlyDeclaredClassesConfusion6` case.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324859](https://bugs.openjdk.org/browse/JDK-8324859): Improve error recovery (**Bug** - P4)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20070/head:pull/20070` \
`$ git checkout pull/20070`

Update a local copy of the PR: \
`$ git checkout pull/20070` \
`$ git pull https://git.openjdk.org/jdk.git pull/20070/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20070`

View PR using the GUI difftool: \
`$ git pr show -t 20070`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20070.diff">https://git.openjdk.org/jdk/pull/20070.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20070#issuecomment-2213597573)

### CODE REVIEW

**Comment by lahodaj** [comment]
FWIW, this is a comparison of errors produced by javac from JDK 17 (`javac-17`) and javac with this patch (`javac-patched`), for a set of inputs:
```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test() // missing open brace
        return true;
    }
    public static boolean test2() {
        return true;
    }
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public static boolean test() // missing open brace
                                ^
TestB.java:6: error: class, interface, enum, or record expected
    public static boolean test2() {
                  ^
TestB.java:8: error: class, interface, enum, or record expected
    }
    ^
3 errors
$ javac-patched TestB.java
TestB.java:3: error: '{' or ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
```

```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test() // missing open brace

    public static boolean test2() {
        return true;
    }
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
$ javac-patched TestB.java
TestB.java:3: error: '{' or ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
```

```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test() // missing open brace
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
$ javac-patched TestB.java
TestB.java:3: error: '{' or ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
```

```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test() // missing open brace
    }
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public static boolean test() // missing open brace
                                ^
TestB.java:5: error: class, interface, enum, or record expected
}
^
2 errors
$ javac-patched TestB.java
TestB.java:3: error: '{' or ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
```

```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test(String,
}
class T {}
$ javac-17 TestB.java
TestB.java:3: error: <identifier> expected
    public static boolean test(String,
                                     ^
TestB.java:4: error: illegal start of type
}
^
2 errors
$ javac-patched TestB.java
TestB.java:3: error: <identifier> expected
    public static boolean test(String,
                                     ^
TestB.java:4: error: illegal start of type
}
^
2 errors
```

```
$ cat TestB.java
package tests;
public class TestB {
    private Object testMethod(final String arg1 final String arg2) {
        return null;
    }
}
$ javac-17 TestB.java
TestB.java:3: error: ',', ')', or '[' expected
    private Object testMethod(final String arg1 final String arg2) {
                                               ^
TestB.java:3: error: ';' expected
    private Object testMethod(final String arg1 final String arg2) {
                                                                 ^
2 errors
$ javac-patched TestB.java
TestB.java:3: error: ',', ')', or '[' expected
    private Object testMethod(final String arg1 final String arg2) {
                                               ^
TestB.java:3: error: ';' expected
    private Object testMethod(final String arg1 final String arg2) {
                                                                 ^
2 errors
```

```
$ cat TestB.java
package tests;
public @interface A {
    public String value() default ""
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public String value() default ""
                                    ^
1 error
$ javac-patched TestB.java
TestB.java:3: error: ';' expected
    public String value() default ""
                                    ^
1 error
```

```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test() // missing open brace
        String s = "";
        return s.isEmpty();
    }
    public static boolean test2() {
        return true;
    }
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public static boolean test() // missing open brace
                                ^
TestB.java:5: error: illegal start of type
        return s.isEmpty();
        ^
TestB.java:5: error: <identifier> expected
        return s.isEmpty();
                        ^
TestB.java:7: error: class, interface, enum, or record expected
    public static boolean test2() {
                  ^
TestB.java:9: error: class, interface, enum, or record expected
    }
    ^
5 errors
$ javac-patched TestB.java
TestB.java:3: error: '{' or ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
```

```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test() // missing open brace
    String s = ""; //field declaration
    public static boolean test2() {
        return true;
    }
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
$ javac-patched TestB.java
TestB.java:3: error: '{' or ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
```

```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test() // missing open brace
        final String s = "";
        return s.isEmpty();
    }
    public static boolean test2() {
        return true;
    }
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public static boolean test() // missing open brace
                                ^
TestB.java:5: error: illegal start of type
        return s.isEmpty();
        ^
TestB.java:5: error: <identifier> expected
        return s.isEmpty();
                        ^
TestB.java:7: error: class, interface, enum, or record expected
    public static boolean test2() {
                  ^
TestB.java:9: error: class, interface, enum, or record expected
    }
    ^
5 errors
$ javac-patched TestB.java
TestB.java:3: error: '{' or ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
```

```
$ cat TestB.java
package tests;
public class TestB {
    public static boolean test() // missing open brace
    final String s = ""; //field declaration?
    public static boolean test2() {
        return true;
    }
}
$ javac-17 TestB.java
TestB.java:3: error: ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
$ javac-patched TestB.java
TestB.java:3: error: '{' or ';' expected
    public static boolean test() // missing open brace
                                ^
1 error
```

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java [hunk 1]
```diff
@@ -5013,14 +5018,69 @@ protected JCTree methodDeclaratorRest(int pos,
...
+                        parseAsBlock = switch (token.kind) {
+                            //definitelly sees a statement:
+                            case CASE, DEFAULT, IF, FOR, WHILE, DO, TRY, SWITCH,
+                                 RETURN, THROW, BREAK, CONTINUE, ELSE, FINALLY,
+                                 CATCH, THIS, SUPER, NEW -> true;
+                            case RBRACE -> {
+                                //check if adding an opening brace would balance out
+                                //the opening and closing braces:
+                                int braceBalance = 1;
+                                VirtualScanner virtualScanner = new VirtualScanner(S);
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:5046]: This seems a new pattern - in other case we use the lookahead method. I don't object to this, as much as note that we now seem to use different styles to lookahead.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java [hunk 2]
```diff
@@ -5013,14 +5018,69 @@ protected JCTree methodDeclaratorRest(int pos,
...
+                    // look if there is a probable missing opening brace,
+                    // and if yes, parse as a block
+                    skip(false, true, !unclosedParameterList, !unclosedParameterList);
+                    boolean parseAsBlock;
                     if (token.kind == LBRACE) {
+                        parseAsBlock = true;
+                    } else if (unclosedParameterList) {
+                        parseAsBlock = false;
+                    } else {
+                        parseAsBlock = switch (token.kind) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:5037]: can we move this "complex" recovery code in a separate method and put a comment in there?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java [hunk 3]
```diff
@@ -5037,6 +5046,84 @@ protected JCTree methodDeclaratorRest(int pos,
...
     }
 
+    /**
+     * After seeing a method header, and not seeing an opening left brace,
+     * attempt to estimate if acting as if the left brace was present and
+     * parsing the upcoming code will get better results than not parsing
+     * the code as a block.
+     *
+     * The estimate is as follows:
+     * - tokens are skipped until member, statement of identifier is found,
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:5056]:
```suggestion
     * - tokens are skipped until member, statement or identifier is found,
```

**Review by mcimadamore — APPROVED** [review]

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**